ee/backend/app/src/main/kotlin/io/tolgee/ee/service/LlmProviderService.kt

```diff
@@ -0,0 +1,297 @@
+package io.tolgee.ee.service
+
+import io.tolgee.component.CurrentDateProvider
+import io.tolgee.component.machineTranslation.TranslationApiRateLimitException
+import io.tolgee.configuration.tolgee.InternalProperties
+import io.tolgee.configuration.tolgee.machineTranslation.LlmProviderInterface
+import io.tolgee.constants.Caches
+import io.tolgee.constants.Message
+import io.tolgee.dtos.LlmParams
+import io.tolgee.dtos.LlmProviderDto
+import io.tolgee.dtos.PromptResult
+import io.tolgee.dtos.request.llmProvider.LlmProviderRequest
+import io.tolgee.dtos.response.prompt.PromptResponseUsageDto
+import io.tolgee.ee.component.llm.*
+import io.tolgee.exceptions.BadRequestException
+import io.tolgee.exceptions.NotFoundException
+import io.tolgee.model.LlmProvider
+import io.tolgee.model.enums.LlmProviderPriority
+import io.tolgee.model.enums.LlmProviderType
+import io.tolgee.repository.LlmProviderRepository
+import io.tolgee.service.LlmPropertiesService
+import io.tolgee.service.organization.OrganizationService
+import org.springframework.boot.web.client.RestTemplateBuilder
+import org.springframework.cache.Cache
+import org.springframework.cache.CacheManager
+import org.springframework.cache.set
+import org.springframework.http.HttpStatus
+import org.springframework.stereotype.Service
+import org.springframework.web.client.HttpClientErrorException
+import org.springframework.web.client.ResourceAccessException
+import org.springframework.web.client.RestTemplate
+import java.time.Duration
+import kotlin.jvm.optionals.getOrNull
+import kotlin.math.roundToInt
+
+@Service
+class LlmProviderService(
+  private val llmProviderRepository: LlmProviderRepository,
+  private val organizationService: OrganizationService,
+  private val llmPropertiesService: LlmPropertiesService,
+  private val openaiApiService: OpenaiApiService,
+  private val tolgeeApiService: TolgeeApiService,
+  private val ollamaApiService: OllamaApiService,
+  private val cacheManager: CacheManager,
+  private val currentDateProvider: CurrentDateProvider,
+  private val claudeApiService: ClaudeApiService,
+  private val geminiApiService: GeminiApiService,
+  private val restTemplateBuilder: RestTemplateBuilder,
+  private val internalProperties: InternalProperties,
+) {
+  private val cache: Cache by lazy { cacheManager.getCache(Caches.LLM_PROVIDERS) }
+  private var lastUsedMap: MutableMap<String, Long> = mutableMapOf()
+
+  fun getProviderByName(
```

Comment:
I am unable to read this method. Some comments what is hapening and why would be useful.